homeassistant/components/asterisk_mbox.py

```diff
@@ -41,42 +44,75 @@ def setup(hass, config):
     port = conf.get(CONF_PORT)
     password = conf.get(CONF_PASSWORD)
 
-    hass.data[DOMAIN] = AsteriskData(hass, host, port, password)
-
-    discovery.load_platform(hass, "mailbox", DOMAIN, {}, config)
+    hass.data[DOMAIN] = AsteriskData(hass, host, port, password, config)
 
     return True
 
 
 class AsteriskData(object):
     """Store Asterisk mailbox data."""
 
-    def __init__(self, hass, host, port, password):
+    def __init__(self, hass, host, port, password, config):
         """Init the Asterisk data object."""
         from asterisk_mbox import Client as asteriskClient
 
         self.hass = hass
+        self.config = config
         self.client = asteriskClient(host, port, password, self.handle_data)
-        self.messages = []
+        self.messages = None
+        self.cdr = None
 
         async_dispatcher_connect(
             self.hass, SIGNAL_MESSAGE_REQUEST, self._request_messages)
+        async_dispatcher_connect(
+            self.hass, SIGNAL_CDR_REQUEST, self._request_cdr)
+        async_dispatcher_connect(
+            self.hass, SIGNAL_DISCOVER_PLATFORM, self._discover_platform)
+
+    @callback
+    def _discover_platform(self, component):
+        discovery.load_platform(self.hass, "mailbox",
+                                component, {}, self.config)
 
     @callback
     def handle_data(self, command, msg):
```

Comment:
Is this called from an async context ? Why does it have the `callback` decorator?